Dima review

KiK0S opened the pull request:
Тут наверняка кучи всего нет, но зато теперь у Димы есть возможность красиво комментировать.

Тем не менее, я запарился достаточно, чтобы про себя считать это версией v0.1

DimaTrushin commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+build/
+*.orig
```
Git по умолчанию ругается на отсутсвие новой строки в конце файла. Эту ругань будет видеть каждый, кто клонирует и форкает твой проект, а так же каждый раз, когда подобный файл будет обновляться у них на машине. Так не стоит делать.

KiK0S replied:
Я не разобрался, как чинить автоматически, но я хотя бы понял, как дрянь выглядит. Вроде поправил

DimaTrushin replied:
Я использую astyle
http://astyle.sourceforge.net/
Это приблуда, которая умеет вносить автоматические правки в код в соответствии со style guide-ами. Она не идеальна и местами делает паршиво, но очень помогает во многих случаях. Такие вещи как новая строка в файлах или удалить все trailing spaces она делает на ура. Плюс я в последнее время QtCreator пользуюсь часто из-за Qt, как можно догадаться, так там эта штука поддерживается IDE на нативном уровне. Я ее настраиваю так, чтобы правки вносились автоматом при каждом сохранении файла.

Есть куча других приблуд, которые позволяют автоматически править код. Но я не большой мастер линтеров и прочей хрени. Сказывается отсутствие опыта работы в 100500 програмистских компаний мира.

KiK0S replied:
> Такие вещи как новая строка в файлах или удалить все trailing spaces она делает на ура.
> Значит, я пока что не нашел нужный флаг. Тогда доразбираюсь, ок

DimaTrushin replied:
Мне казалось, что новую строчку в конце файла он автоматом делает, как и trailing spaces удаляет без дополнительных флагов.

DimaTrushin commented on src/camera.h:
```diff
@@ -0,0 +1,35 @@
+public:
+    Camera() = delete;
+    Camera(Screen& screen);
+    Camera& operator = (const Camera& camera);
```
У любого класса есть несколько специальных методов, которые компилятор генерит автоматически (либо перестает генерить автоматическиа, делая их необъявленными), это:
1) default constructor
2) copy constructor
3) move constructor
4) copy assignment
5) move assignment
6) destructor

Теперь, есть несколько правил, которые надо соблюдать. Copy constructor и copy assignment надо определять парой одновременно. Аналогично с move операторами. Иначе UB вообще говоря. У тебя определен copy assignment operator, но при этом не определен copy constructor. Надо добавить второй. При этом можно написать имплементацию assignment operator через соответствующий constructor. 

Подробнее можно поглядеть тут
https://youtu.be/vLinb2fgkHk?t=818
Я тайминг выставил на объяснении специальных методов и правил их соглаосвания. Но до этого есть описание стандартных имплементаций всех этих товарищей, потому полезно посмотреть само видео целиком.

KiK0S replied:
Мне понравилось rule of zero, попытаюсь придерживаться его

DimaTrushin commented on src/camera.h:
```diff
@@ -0,0 +1,35 @@
+    Matrix4d inverse() const;
+
+private:
+    Screen& screen_;
```
Если тебе нужно иметь ссылку на другой объект, так чтобы этот другой объект можно было менять, то обычно используют указатель, а не ссылку. Ссылки обычно используют константные. Проблема со ссылками в том, что у них очень мерзкая семантика, которая приводит к созданию костылей в виде написания каких-то кастомных конструкторов и assignment operators и так далее. Невладеющий указатель, в свою очередь, ведет себя в соответствии с value semantics всегда. Кроме того, ссылку ты не можешь переопределить, а указатель можешь. В итоге у тебя будет очень странное поведение у опирации копирования, когда ты пытаешься скопировать камеру с одним экраном в камеру с другим, а экран в итоге не копируется.
Единственное, что надо будет добавить при использования указателей -- поставить везде assert-ы на проверку, что указатель не nullptr везде, где ты к нему обращаешься.

KiK0S replied:
Понял про указатели, разумно. А зачем покрыть все ассертами, если это приватное поле?

DimaTrushin replied:
Смотри, бывает два типа ошибок в коде:
1) Ошибка вида -- Я баклан написал говно.
2) Я вызвал функцию из api операционки и она упала.
- Во втором случае ты бессилен и для этого придумали exception.
- В первом случае в твоих интересах все такие ошибки поправить. Главная идея исправления ошибок -- программа должна упасть, как только встретила ошибку и сообщить тебе об этом. Падение программы в debug -- это хорошо. Это помогает тебе находить твои ошибки. Для этого и нужен assert. Он в debug режиме валит твою прогу, если не выполнены условия, а в release он вырезается и не дает кода, а потому не портит производительность.
Правильная идея использования assert такая. Если у тебя есть класс, то обычно внутренние данные класса должны удовлетворять каким-то инвариантам. Или если у тебя есть функция, то часто подразумевается, что аргументы удовлетворяют некоторым условиям. В подобных случаях ты можешь проверять assert-ами такие инварианты. И если они сломались, ты можешь отловить проблему как можно ближе к куску кода, где ошибка (это может быть банальная опечатка по невнимательности, но ты ее хрен найдешь иначе, а без assert-ов не заметишь, пока не упадет у клиента).
Пример. Пусть у тебя есть класс с указателем, который ты хочешь разыменовывать. Например
class B{
public:
void func() {}
};
class A{
public:
B* operator->() const {
return pValue_;
}
bool isDefined() const {
return pValue_ != nullptr;
}
private:
B* pValue_;
};
Логика использования класса такая: ты проверяешь определен ли объект и если определен, то имеешь право его разыменовать.
A a;
if (a.isDefined())
  a->func();
Но что если ты забыл проверить условие и просто вызвал оператор ->? В этом случае программа упадет в run-time. А может и не упадет, потому что указатель будет не нулевым из-за того, что по этому адресу находится мусор или еще в силу каких трюков оптимизации или еще чего можно встретить в релизе. Но раз ты вызвал оператор без проверки, это явно твоя ошибка как программиста, которую надо исправить. Потому тебе компилятор должен помочь найти эту ошибку. Тогда ты можешь сделать так
B* operator->() const {
assert(isDefined());
return pValue_;
}
Ты предполагаешь, что когда вызываешь этот оператор, то pValue_ обязательно не nullptr. Вот мы и проверяем этот инвариант. Если вдруг, ты где-то в другом месте сделал что-то не так и этот инвариант нарушился, то у тебя программа упадет в этом месте и ты сможешь отслеживать ошибку.

KiK0S replied:
Ладно, справедливо. Просто все еще живу с уверенностью, что пишу код без багов, хехе.

DimaTrushin replied:
Запомни главную аксиому:
"В моем коде всегда есть баги"
Важен процент багов в написнном коде. С опытом этот процент сильно падает. И даже если ты перестал делать ошибки в коде, то это не значит, что их не появится чисто из статистических соображений. Особенно они любят появляться, когда торопишься и делаешь вещи, которые понятно как сделать и перестаешь быть внимательным.

KiK0S replied:
Понял, покрою ассертами и не буду выделываться

DimaTrushin commented on src/camera.h:
```diff
@@ -0,0 +1,35 @@
+    Camera& operator = (const Camera& camera);
+
+    template <typename T>
+    sf::Vector2f project_point(Point4d<T> p) const;
```
Шаблонный метод нельзя определять в header, а объявлять в source файле. Это ошибка линковки. Тут надо знать модель компиляции c++. Если кратко, то у тебя работают компилятор и линкер. 
1) Сначала компилятор проходит по cpp файлам и компилирует все функции(!!!) которые он видит и складывает результат в промежуточные бинарные файлы. При этом компилятор не должен знать определение всех функций, чтобы скомпилировать единицу трансляции. Если при компиляции встречается вызов функции и ее определение не видно компилятору, то он в бинарнике оставляет точку вызова функции, которую надо будет разрешить во время линковки.
2) Потом работает линкер. Он находит точку входа в программу main. Идет по ее телу, если там нет вызова неизвестных функций, то весь исполняемый код найден, его складываем в exe и радуемся. Если есть неопределенные функции, то ищем их в других бинарниках. И если не нашлось ни одной, то это ошибка линковки, если нашлось несколько и они не помечены специальным флагом inline, то это ошибка линковки.

Теперь, когда ты работаешь с шаблонами, то надо понимать, что шаблон -- это НЕ функция, это blueprint функции. То есть шаблон -- это указания компилятору, как собрать функцию при известных параметрах шаблона, а не сама функция. И компилятор не будет ее генерировать до тех пор, пока не встретит где-то использование этого шаблона. Но для того, чтобы сгенерировать, компилятору надо видеть полное определение шаблона.

Пример
Пусть у тебя есть три файла
some.h
some.cpp
main.cpp
Теперь
some.h
void f();
template<class T >
void g();
some.cpp
#include "some.h"
void f() {}
template<class T >
void g() {}
main.cpp
#include "some.h"
int main() {
f();
g<int >();
}
Далее у тебя есть 2 cpp файла. Это значит, что компилятор отдельно скомпилирует some.cpp и отдельно main.cpp и создаст some.o и main.o соответственно.
Давай посмотрим, что произойдет при компиляции some.cpp
#include директива -- это просто копирование текста.
В итоге компилятор видит, что у него есть объявление функции
void f();
и сама функция
void f() {}
он скомпилирует тело функции и положит его в some.o
Потом компилятор видит, что есть объявление шаблона
template<class T >
void g();
и определение шаблона
template<class T >
void g() {}
Но так как это не конкретная функция, а всего лишь шаблон, то компилятор не может догадаться с каким шаблонным параметром надо скомпилировать этот шаблон и просто пропускает это определение. В итоге в файле some.o лежит лишь тело функции f()
Далее компилятор будет компилировать main.cpp
Он вставляет содержимое some.h в начале. Это дает понять компилятору, что имена f и g являются функцией и шаблоном функции соответственно. Теперь компилятор находит main и начинает компилировать строчка за строчкой
Он видит имя f, понимает, что это функции и что у него нет в этой единице трансляции тела функции, потому в этом месте он оставляет точку вызова фукнции f.
Далее он видит g<int >() и понимает, что тут его просят взять шаблон g и инстантиировать его с помощью класса int и вызвать соответствующую функцию. Так как компилятор не видит определения шаблона, а видит лишь его объявление, то он опять же оставляте тут точку вызова данного шаблона. В итоге в main.o лежит функция main в которой в начале идет точка вызова f, потом точка вызова g<int >.

Теперь компиляция закончена, работает линкер. Линкер берет main.o и находит там точку входа main. Далее он смотрит, что есть вызов f. Он находит f в файле some.o и вставляет код функции f в точку вызова f. Далее он видит, что есть вызов g<int >, начинает искать эту функцию где-то в других *.o файлах и не находит ее. Ошибка линковки.


В сухом остатке. С шаблонами тебе надо обязательно определение функций помещать в header. Либо делать inline в теле класса, либо вне класса, но в конце header файла.

KiK0S replied:
у меня подгорело с этого факта. Ну, избавлюсь от шаблонов везде, я не гордый.

DimaTrushin replied:
Главный вопрос в том, нужны ли тебе шаблонные функции именно в этом месте? Раз это твой класс, то тебе решать с чем ты хочешь работать с double или с произвольным типом T. Но когда ты работаешь с произвольным типом T не забывай, что он может оказаться char* или еще каким говном похуже, и самое отвратное, что он может конвертироваться в double или float без каких-либо предупреждений. Это вопрос дизайна. Почему вдруг у тебя везде все функции для double, а именно здесь понадобились шаблоны? Это выглядит подозрительно. Шаблоны -- очень хорошая и нужная вещь, но как любой молоток, им надо забивать гвозди, а не полировать стекло или варить суп. Я так вообще без шаблонов никуда. Но они обычно являются частью библиотечного кода, когда мне нужно что-то очень универсальное, что потом можно использовать кучу раз. Тут был больше поинт в том, чтобы ты подумал, а чего ты тут хочешь сделать.

KiK0S replied:
Ну в целом я бы сказал, что маловероятно, что я где-то хочу по-настоящему использовать другие типы данных. То есть, иногда бывают приколы с разным подбором числовых типов, но поскольку тут все крутится-вертится, то наверное не очень хочется.

Если подумать о том, буду ли я переиспользовать этот код в рамках других проектов — почти гарантированно нет. Если отказ от шаблонов типа принесет мне однородную структуру проекта и отсутствие головной боли, то может быть действительно есть смысл решить, что сейчас я кухарка, а не строитель.

Думаю, сейчас попытаюсь как-то логично отказаться от шаблонного усложнения кода

DimaTrushin replied:
Кстати, заметил, что на гит плохо отображаются открывающиеся и закрывающиеся скобки, если не поставить нужное количество пробелов. Я вроде поставил, но не туда куда надо. В итоге весь код с шаблонами отобразился непойми как. Надеюсь от этого пример не стал менее понятным. Я только сейчас это увидел.

DimaTrushin commented on src/camera.h:
```diff
@@ -0,0 +1,35 @@
+
+class Camera {
+public:
+    Camera() = delete;
```
Тебе не обязательно удалять дефолтный окнструктор. Как только ты объявил свой конструктор (любой, в данном случае Camera(Screen& screen) ), то компилятор не будет генерировать дефолтный конструктор самостоятельно. Это так же описано в в том видео про специальные методы класса, что я кидал в другом комментарии.

KiK0S replied:
Да, я уже посмотрел и осознал, спасибо!

DimaTrushin commented on src/camera.h:
```diff
@@ -0,0 +1,35 @@
+    Camera& operator = (const Camera& camera);
+
+    template <typename T>
+    sf::Vector2f project_point(Point4d<T> p) const;
```
- Еще эта функция выглядит как деталь имплементации. Не понятно почему она должна быть у камеры в открытом интерфейсе. Вообще не понятна роль камеры.
- Кроме того, не понятно почему этот метод должен быть шаблонным. Неужели у тебя бывает что-то кроме double? Но даже если бывает, то не понятно, почему тогда Screen, Matrix и прочие компоненты не зависят от этого параметра? Куда разумнее было бы тогда сделать камеру целиком шаблонную заточенную под другой тип. Но если кроме double ты ничем не пользуешь, то в этом нет смысла. Если пользуешь float, то у тебя всегда есть неявная конвертация float к double.

Вообщем это выглядит, как ошибка дизайна.

KiK0S replied:
про второе понятно, про первое -- хз, пока что камера на очень полезна, но предполагается, что это обертка, которая потом будет хранить данные про всякие фокальные точки и быть оберткой над матрицей преобразований.

DimaTrushin replied:
Я понимаю, что так как это первая попытка написать, у тебя в голове нет целостной картины, что за что отвечает и какие есть компоненты. Это нормально. Такая ситуация будет всегда, когда делаешь что-то, с чем не встречался. Но, я хочу тут сделать одно важное замечание. Смотри, как ты описываешь роль камеры:
" это обертка, которая потом будет хранить данные про всякие фокальные точки и быть оберткой над матрицей преобразований."
То есть ты описываешь ее имплементацию, что ты в ней будешь хранить, а не что она будет делать! Это ошибка мыслительного процесса. В объекте имплементация имеет второстепенное значение, куда важнее, что этот объект будет делать (это его открытый интерфейс). То есть тебе правильно задаться вопросом не "что камера будет хранить", а "что камера будет делать". Например, что ты камере можешь скормить точку в пространстве и экран, она умеет спроектировать ее на экран (она может быть и хранит экран в себе как объект, но про это лучше думать так, что камера просто знает про существование экрана и вообще может не хранит, например, через указатель). Вместо того, чтобы думать, что камера хранит фокальную точку, лучше думать, что камера знает как найти фокальную точку (например у нее есть метод для этого), а хранит она ее или берет откуда-то из другого места или высчитывает по хитрому алгоритму -- это второстепенные вопросы. Вот это правильный подход думать, что объект делает и на каких этапах, какие ему нужны данные для каждой операции, откуда он берет эти данные. Вот эти вопросы помогут тебе придумывать хороший дизайн и поправлять (рефакторить) плохой дизайн.

DimaTrushin commented on src/camera.h:
```diff
@@ -0,0 +1,35 @@
+    template <typename T>
+    double get_z_value(Point4d<T> p) const;
+
+    void apply_matrix(Matrix4d matrix);
```
Мне как пользователю камеры не понятно, что означает эта функция с точки зрения камеры. Что значит применить матрицу к камере? Смысл этой матрицы? Это матрица преобразования над камерой? Если да, то в каких координатах? Если в локальных, то лучше выделить отдельные виды преобрзований вроде вращения вдоль такой-то оси и так далее, если в глобальных, то это значит, что ты применяешь трансформацию к камере как к объекту 3d сцены и тут я бы ожидал другое название метода.

DimaTrushin commented on src/camera.h:
```diff
@@ -0,0 +1,35 @@
+    double get_z_value(Point4d<T> p) const;
+
+    void apply_matrix(Matrix4d matrix);
+    Matrix4d inverse() const;
```
Вот это вообще не понятно, что за метод такой у камеры. У меня даже гипотез нет, что этот метод делает.

DimaTrushin commented on src/camera.h:
```diff
@@ -0,0 +1,35 @@
+
+private:
+    Screen& screen_;
+    Point4d<> camera_;
```
Очень плохое название для переменной. В объекте типа Camera у тебя есть переменная camera_. Это выглядит как рекурсия. Это плохо. Что эта точка делает? Это центр экрана? Это фокальная точка камеры? Назови объект своим именем.

DimaTrushin commented on src/camera.h:
```diff
@@ -0,0 +1,35 @@
+private:
+    Screen& screen_;
+    Point4d<> camera_;
+    Matrix4d transform_;
```
Вот это не понятно, что за transform. Какая роль у этой матрицы? Это матрица для делания чего? Это проекция на плоскость экрана камеры? Это проективное преобразование которое переводит пирамиду зрения в куб? Назови преобьразование тем именем, которое у него должно быть.

DimaTrushin commented on src/camera.h:
```diff
@@ -0,0 +1,35 @@
+    Point4d<> camera_;
+    Matrix4d transform_;
+    double MAX_Z = 100000;
+    double MIN_Z = -100000;
```
- Вот тут я не совсем понимаю назначение этих параметров. Их можно вообще настраивать или это должны быть константы? Это ограничение области видимости камеры вдаль? Если так, то почему у тебя отрицательный параметр? Где находится по умолчанию плоскость экрана?
- Тут одна из двух проблем:
-- либо неправильно выбраны параметры для камеры MAX_Z и MIN_Z и к ним присвоили магические константы, 
-- либо ты объявил константы не сказав, что это константы.
В первом случае я бы ожила что-то такое
double MAX_Z = k_max_view_depth;
и перед этим объявить
static constexpr const double k_max_view_depth = 100000;
Это устраняет магические константы из кода.
Если у тебя второй случай, то надо сразу писать
static constexpr const double MAX_Z = 100000;
но тут плохое имя константы, и не понятна ее роль.

DimaTrushin commented on src/camera.h:
```diff
@@ -0,0 +1,35 @@
+    Screen& screen_;
+    Point4d<> camera_;
+    Matrix4d transform_;
+    double MAX_Z = 100000;
```
Непонятно, почему эти переменные никак невозможно модифицировать (нельзя задать в конструкторе, нет модификаторов для них, нет методов доступа) и при этом они не const double. А если их нельзя модифицировать, то они одинаковы для всех instance твоего класса, тогда не понятно, почему они не static const double. Но в этом случае они должны быть constexpr. Вообщем, не понятно, что тут происходит. Так же смотри другой комментарий к следующей переменной MIN_Z.

DimaTrushin commented on src/camera.cpp:
```diff
@@ -0,0 +1,46 @@
+
+Camera::Camera(Screen& screen): screen_(screen),
+    camera_(Point4d<double>(0, 0, 0)) {
+    transform_[0][0] = 1;
```
члены класса надо инициализировать в списке инициализации. Такая инициализация означает, что у Matrix4d не хватает нужного конструктора. Например можно было бы сделать
class Matrix4d {
...
public:
static Matrix4d IdentityMatrix();
private:
Matrix4d(double a11, double a12, ...);
...
};
Matrix4d Matrix4d::IdentityMatrix() {
return Matrix4d(1,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1);
}

Используется так
Matrix4d Mat = Matrix4d::IdentityMatrix();
При этом задействуется rvo и объект единичной матрицы создается сразу по адресу Mat.

Что такое rvo можно почитать тут (это текст из другого ревью, проигнорируй нерелевантные вещи, я просто задолбался писать одно и то же и все никак не могу начать писать guide по плюсам):
кроме copy и move есть еще такая вещь как rvo или copy elision. Вот тут можно почитать как оно работает (это актульно для until c++17, но помогает понять, что происходит):
https://www.ibm.com/developerworks/community/blogs/5894415f-be62-4bc0-81c5-3956e82276f3/entry/RVO_V_S_std_move?lang=en
Еще хорошее объяснение есть вот тут
https://stackoverflow.com/questions/12953127/what-are-copy-elision-and-return-value-optimization
и тут
https://stackoverflow.com/questions/38043319/how-does-guaranteed-copy-elision-work
По-хорошему тут надо писать так:
Monomial tmp = *this;
tmp *= x;
return tmp;
В этом случае сработает rvo. Так как у вас локальный объект, который не является аргументов функции и тип возвращаемого объекта в точности совпадает с типом возвращаемым функцией. Наличие std::move меняет тип возвращаемого объекта и мешает rvo. Про это хорошо написано по первой ссылке.
В c++17 во время возврата значения из функции вообще не создается временной переменной:
https://en.cppreference.com/w/cpp/language/copy_elision

KiK0S replied:
Да, я сделал нужный конструктор теперь

DimaTrushin commented on src/camera.cpp:
```diff
@@ -0,0 +1,46 @@
+Camera& Camera::operator = (const Camera& camera) {
+    camera_ = camera.camera_;
+    transform_ = camera.transform_;
+    return (*this);
```
Убрать скобки вокруг *this
Плюсы устроены так, что лучше не использовать лишние скобки, они могут поменять контекст. Потому не используй эту дурацкую привычку. Тут разницы нет, но просто не надо.

DimaTrushin commented on src/camera.cpp:
```diff
@@ -0,0 +1,46 @@
+    transform_[3][3] = 1;
+}
+
+Camera& Camera::operator = (const Camera& camera) {
```
- Дизайн копирования странный, у тебя screen не меняется при копировании. Что плохо и нарушает value semantics. Это может жестко аукнуться при оптимизациях компилятора, если screen в this и в camera разные.
- Еще название переменной. Лучше писать other вместо camera ибо тут и так понятно, что ты работаешь с представителем класса Camera это раз. Во-вторых, имя camera не понятно что значит, какая именно камера? Твоя текущая или другая или вообще какой-то другой тайный смысл?
- Кроме того, ты не копируешь MAX_Z и MIN_Z, потому мне начинает казаться, что это статические константы класса, а не переменные объекта класса.

Что надо сделать:
1) сделать screen указателем
2) сделать MAX_Z и MIN_Z constexpr константами
3) После этих модификаций, компилятор для тебя сгенерирует cope constructor и copy assignment по умолчанию такими какими надо и не надо будет писать их явно.
Единственное, что в конструкторе через Screen* надо будет поставить assert проверяющий, что указатель не nullptr.

DimaTrushin commented on src/camera.cpp:
```diff
@@ -0,0 +1,46 @@
+
+void Camera::apply_matrix(Matrix4d matrix) {
+    camera_ = matrix * camera_;
+    transform_ = matrix* transform_;
```
Судя по тому, что я вижу, матрица transform_ хранит данные локального репера для камеры. Это надо отразить в названии этой матрицы. Назначение camera_ мне пока не ясно никак.

DimaTrushin commented on src/camera.cpp:
```diff
@@ -0,0 +1,46 @@
+    transform_ = matrix* transform_;
+}
+
+Matrix4d Camera::inverse() const {
```
Похоже на ненужный метод. Камера не матрицы должна обращать, а делать более высокоуровневую работу. Явно тебе эта матрица нужна для чего-то, что должна делать камера, а не просто так, как конечный результат работы камеры.

DimaTrushin commented on src/camera.cpp:
```diff
@@ -0,0 +1,46 @@
+template <typename T>
+sf::Vector2f Camera::project_point(Point4d<T> p) const {
+    p = transform_ * p;
+    if (p.z / p.w + screen_.SCREEN_SIZE / 2 > MIN_Z && p.z / p.w + screen_.SCREEN_SIZE / 2 < MAX_Z) {
```
Когда у тебя такие длинные условия, то надо их выделять в одну отдельную функцию. Сравни это со следующей строчкой
if (isVisible(p))
Я не уверен, что я правильно понимаю, что это условие делает, но выделение отдельной функции с говорящим имененм -- намного лучше, чем писать длинные строчки со страшными условиями. Потому что компилятор все равно это заинлайнит и в производительности разницы не будет, а человеку проще прочитать имя функции, что она делает, а не пробираться сквозь дебри логических операторов. Если я захочу проверить корректность работы функции, мне надо будет лишь один раз разобраться в том, что ты тут делаешь.

DimaTrushin commented on src/camera.cpp:
```diff
@@ -0,0 +1,46 @@
+    if (p.z / p.w + screen_.SCREEN_SIZE / 2 > MIN_Z && p.z / p.w + screen_.SCREEN_SIZE / 2 < MAX_Z) {
+        return {p.x / p.w, p.y / p.w};
+    }
+    return {0, 0};
```
Магическая константа. Что это за точка с координатами (0, 0)? Это заглушка, просто какая-то точка? Это центр экрана? Какой смысл у нее? Лучше написать в закрытой части класса камеры что-то в таком духе:
static constexpr const sf::Vector2f kSomeDefaultPoint = sf::Vector2f(0,0);
а потом уже тут использовать
return kSomeDefaultPoint;
Имя надо подобрать подобающим!
Это общая стратегия борьбы с магическими константами. Ты добавляешь статическую constexpr константу в нужную область видимости.

DimaTrushin commented on src/camera.cpp:
```diff
@@ -0,0 +1,46 @@
+    if (p.z / p.w + screen_.SCREEN_SIZE / 2 > MIN_Z && p.z / p.w + screen_.SCREEN_SIZE / 2 < MAX_Z) {
+        return p.z / p.w;
+    }
+    return screen_.SCREEN_SIZE * 10;
```
Вот это вообще что за феерия? 
- Во-первых, магическая константа 10.
- Во-вторых, что вообще тут вычисляется?? Десять размеров экрана -- это z value???

DimaTrushin commented on src/chair.h:
```diff
@@ -0,0 +1,17 @@
+class Chair : public SurfaceObject<double> {
+public:
+    Chair();
+    ~Chair();
```
Непонятно зачем тебе явно объявлять деструктор, который ничего не делает. Если это надо из-за того, что ты наследуешься от SurfaceObject<double>, то можно написать
~Chair() = default;
Еще сразу вопрос. Ты обязан наследоваться от surfaceObject<double> именно public?

DimaTrushin commented on src/chair.h:
```diff
@@ -0,0 +1,17 @@
+    ~Chair();
+
+private:
+    const int SIZE = 100;
```
static constexpr const int kSize_ = 100;
- Во-первых, static, чтобы у всех instance класса это была одна и та же переменная
- Во-вторых, constexpr, чтобы компилятор заинлайних это значение в код на этапе компиляции.
- В-третьих, НЕ используй заглавные буквы ни для чего кроме макросов.
- В-четвертых, константы обычно снабжаются префиксом k в начале.
- В-пятых, закрытые данные класса надо снабжать underscore в конце, чтобы отличать эти переменные от локальных переменных методов класса.

Я еще посмотрел, у тебя эта переменная играет роль масштаба, а не размера, то есть scale было бы более подходящим.

KiK0S replied:
насчет имени scale хз, потому что не воспринимаю это как масштаб. мб проблема восприятия. остальное разумно очень.

DimaTrushin replied:
Имя выбирай как тебе больше кажется разумным. Это был не принципиальный момент.

DimaTrushin commented on src/chair.cpp:
```diff
@@ -0,0 +1,39 @@
+namespace app {
+
+Chair::Chair() {
+    WireObject<double>::points_ = {
```
Неужели у SurfaceObject<double> от которого ты наследуешься нет соответствующего конструктора, чтобы делегировать ему эту работу в списке инициализации? Это может быть ошибка дизайна библиотеки.

DimaTrushin commented on src/classes.h:
```diff
@@ -0,0 +1,11 @@
+#pragma once
```
Какая роль у этого файла??

KiK0S replied:
Ну сделаем вид, что его больше нет. Раньше все ломалось, когда хранились ссылки друг на друга. С указателями должно работать.

KiK0S replied:
Не, все еще не работает, пришоось его вернуть. По сути, это forward declaration всего

DimaTrushin replied:
Сделай forward declarations там где это надо только в тех header-ах, где надо и только тех классов, которые нужны.

DimaTrushin commented on src/classes.h:
```diff
@@ -0,0 +1,11 @@
+class Screen;
+class Frame;
+class Camera;
+template<size_t N, size_t M> class Matrix;
```
Вот это бесполезная строчка. Я уже объяснял про шаблоны функций. Тут будет та же петрушка, только еще мозгодробительнее.

KiK0S replied:
Ты уже объяснил, я понял не до конца, но ок. Примерно понятна идея. Избавлюсь от файла и сделаю матрицу нешаблонной

DimaTrushin replied:
Я не имел в виду, что матрицу надо делать не шаблонной. Я просто хочу сказать, что шаблон должен целиком находиться в header файле и объявления и определения. То есть
some.h
template<class T>
class A {
public:
void func();
};

template<class T>
void A<T>::func() {
}

Иначе будет ошибка линковки (судя по тому, что у тебя работает, ты, видимо, где-то cpp файлы инклудишь). У тебя же здесь идет только forward declaration. Зачем все это нужно в одной куче не понятно.

DimaTrushin commented on src/cube.h:
```diff
@@ -0,0 +1,16 @@
+template <typename T = double>
+class Cube : public WireObject<T> {
+public:
+    Cube();
```
Так как класс шаблонный, то конструктор надо делать inline иначе будет ошибка линковки. Либо не мучить жопу и сделать все свои классы рассчитанными на double. При передачи double в sfml у тебя произойдет конвертация double к float где надо. Только в этом месте надо заботиться о конвертации. Это сделает дизайн твоих классов более consistent.

KiK0S replied:
Не буду мучить жопу.

DimaTrushin commented on src/cube.h:
```diff
@@ -0,0 +1,16 @@
+    Cube();
+
+private:
+    const double SIZE = 200;
```
аналогичное замечание что было выше.

DimaTrushin commented on src/cube.cpp:
```diff
@@ -0,0 +1,28 @@
+
+template <typename T>
+Cube<T>::Cube() {
+    WireObject<T>::points_ = {
```
Аналогичное замечание про делегирование конструктора.

DimaTrushin commented on src/matrix.h:
```diff
@@ -0,0 +1,32 @@
+    Matrix<N, N> inverse() const ;
+
+private:
+    std::vector<std::vector<double>> data_;
```
Очень плохая имплементация.
- Вектор векторов не лежит в едином сегменте памяти. Надо использовать один вектор размера M * N.
- Ошибка дизайна. У тебя размеры M и N для матрицы являются константами компилятции, а размер вектора является динамическим. Я бы на твоим месте использовал std::array<double, M * N>. Это сделает Matrix не move friendly, но зато все будет лежать в одной области памяти. То есть тебе надо будет заботиться о том, чтобы матрицы никогда не передавались by value, но сильно ускорит работу с матрицами (на порядки). А для малых матриц вроде 4 на 4 или 3 на 3 будет просто феерический выигрыш в производительности.

DimaTrushin commented on src/matrix.h:
```diff
@@ -0,0 +1,32 @@
+    Matrix<N, K> operator * (Matrix<M, K> other) const ;
+    Vector4d operator *(Vector4d other) const ;
+
+    Matrix<N, N> inverse() const ;
```
здесь надо enable_if добавить проверяющий, что M = N. чтобы отключать этот метод в случае, если M не равно N.
Про то, как это работает на примере конструкторов можно посмотреть тут
https://www.youtube.com/watch?v=PNRju6_yn3o
Это вообще очень полезное видео про move семантику и как правильно ей пользоваться, какие там есть подводные камни.

DimaTrushin commented on test/test.cpp:
```diff
@@ -0,0 +1,80 @@
+#include <cmath>
+
+
+int main() {
```
Отвратительно длинная main функция. Философия написания main такая, в ней три строчки:
- считать данные,
- обработать данные,
- вывесит данные.
Ну или в случае интерактивного приложения: 
- Создать все ресурсы для работы приложения
- Запустить event loop
Все
Вся остальная мелкоуровневая работа должна выполняться в другом месте. Это все по-хорошему надо отрефакторить.

DimaTrushin reviewed (changes requested):
Я прошелся по части файлов и накидал комментарии по первому, что приходит на ум. Сейчас уже не хватило сил пробраться через все. Я не комментил одинаковые проблемы, потому надо самому по аналогии находить похожие вещи и править. Кинул пару ссылок на полезные ресурсы посмотреть и почитать, думаю, что пригодятся.
В целом оформление кода очень приличное, но дизайн пока очень сильно хромает (это нормально). Есть разные проблемы с плюсами, которые надо просто будет запомнить и выучить (тут нет другого выхода, как напарываться на эти грабли, узнавать про них и править).

DimaTrushin commented on src/camera.h:
```diff
@@ -2,31 +2,26 @@
-    Screen& screen_;
-    Point4d<> camera_;
+    Screen* screen_ = nullptr;
+    Point4d camera_;
```
- Ты можешь написать
Point4d camera_ = Point4d(0, 0, 0);
Matrix4d transform_ = Matrix::identity_matrix();
И тогда в любом конструкторе в списке инициализации эти поля будут автоматически инициализироваться этими значениями.
- Имена. camera_ -- это что за точка у камеры? Может быть camera_position_?
- Аналогично transform_, но про это уже есть замечание.

KiK0S replied:
Про имена я поменяю, как разберусь кого чем хочу называть!

KiK0S replied:
>И тогда в любом конструкторе в списке инициализации эти поля будут автоматически инициализироваться этими значениями.

DimaTrushin commented on src/camera.cpp:
```diff
@@ -1,46 +1,37 @@
-}
+Camera::Camera(Screen* screen):
+    screen_(screen),
+    camera_(Point4d(0, 0, 0)),
```
Если в header-е написать инициализацию как я указал, то тут можно убрать camera_ и transform_  из списка инициализации, они будут инициализироваться автоматически.

DimaTrushin commented on src/camera.cpp:
```diff
@@ -1,46 +1,37 @@
+Camera::Camera(Screen* screen):
+    screen_(screen),
+    camera_(Point4d(0, 0, 0)),
+    transform_(Matrix::identity_matrix()) {}
```
Добавь
assert(screen_);
чтобы гарантировать, что пользователь нам не поднасрал и не выдал nullptr. Это общее правило, всегда, когда ты скармливаешь функции указатель в качестве аргумента, тут же на автомате ставь assert-ы на проверку того, что эти указатели не нулевые.

KiK0S replied:
Здравая мысль, спасибо!

DimaTrushin commented on src/camera.h:
```diff
@@ -2,31 +2,26 @@
+    sf::Vector2f project_point(Point4d p) const;
+    double get_z_value(Point4d p) const;
 
     void apply_matrix(Matrix4d matrix);
```
Matrix4d достаточно тяжелый объект. Теперь, когда ты хранишь std::array<double, 16> внутри, при передаче by value, ты копируешь 16 double-ов. Тут надо тестить, что будет быстрее -- передавать по константной ссылке или по значению. Предлагаю потестить другую сигнатуру
void apply_matrix(const Matrix4d& matrix);
может быть будет быстрее, может медленнее, зависит от разных факторов. Но стоит иметь в виду, особенно если ты запариваешься по производительности. (Это не самая фатальная дыра, что я нашел).

DimaTrushin commented on src/camera.h:
```diff
@@ -2,31 +2,26 @@
-    sf::Vector2f project_point(Point4d<T> p) const;
-    template <typename T>
-    double get_z_value(Point4d<T> p) const;
+    sf::Vector2f project_point(Point4d p) const;
```
Аналогичное замечание было про Matrix4d против const Matrix4d&. Структура Point4d уже более легкая и для нее by value действительно может быть более эффективно, чем by reference. На всякий случай озвучу следующие правила:
- встроенные типы надо всегда передавать by value. Например чистые int, double, float и т.д.
- свои типы почти всегда надо передавать by const reference. Исключение -- структуры, которые по размеру близки к встроенным типам. 

В сухом остатке. Здесь думаю, стоит оставить как есть. А потом, когда будет нечем поиграться и посмотреть будет ли разница.

DimaTrushin commented on src/camera.h:
```diff
@@ -2,31 +2,26 @@
 
 #include <SFML/Graphics.hpp>
```
Обычно include делают в следующем порядке:
1) свои include в Алфавитном порядке
2) include 3 party library в Алфавитном порядке
3) include из STL в Алфавитном порядке

Есть принцип наоборот от самых общих к самым частным. Но у этого подхода есть такое преимущество. Если ты свои делаешь раньше, то компилятор ругнется, в случае если ты забыл в своем header что-то подключить из более общего. Все что помогает уронить программу из-за твоих ошибок считается хорошей практикой.

DimaTrushin commented on src/frame.h:
```diff
@@ -2,35 +2,33 @@
     std::vector<std::vector<sf::Color>> color_;
     std::vector<std::vector<double>> z_value_;
-    std::vector<WireObject<> *> objects_;
+    std::vector<WireObject*> objects_;
```
На сколько я вижу у тебя объекты сцены хранятся где-то, а тут ты хранишь указатели на  них. Это плохая идея, очень плохая. Если объекты хранятся в контейнере, у которого указатели на объекты могут измениться (например std::vector, который реаллоцирует), то у тебя сразу все повалится и ты хрен отследишь почему.
В плюсах надо постоянно думать про владение объектов, где они живут и какая у них семантика. Самая безопасная семантика -- value semantics. К ней относится передача объектов by value и by const shared reference. Не буду углубляться в детали тут, но это отдельный пласт философии плюсов. Всегда надо обращать внимание на объекты типа reference и pointer. С первыми работать паршиво, а вторые могут начать указывать не туда, куда надо. Обычно объекты, которые передаются по указателю, должны подчиняться специальным правилам вроде "быть адресуемыми". К адресуемым объектам например относятся те, что создаются на heap или подъобъекты адресуемых.

В итоге. Я бы сказал, что тут надо по-другому организовывать работу класса. Если бы я хранил объекты в контейнере Objects где-то в программе, то сюда я бы передавал целиком весь контейнер, точнее константную ссылку на него, то есть что-то вроде
void draw(const ObjectsContainer&);
Это избавляет тебя от необходимости хранить внутри указатели на индивидуальные объекты, которые могут поехать. Вообще этот метод должен быть у Renderer. Я сейчас плохо понимаю происходящее тут.

KiK0S replied:
Блин, я попозже про это напишу понятно, тут сложная схема (а еще я потом перечитаю, что ты тут написал, я не во всем разобрался). Основная причина, по которой я сделал так — я потом различаю поведение для WireObject и SurfaceObject:WireObject. Для одного рисую только ребра, для другого и ребра и грани. Поэтому у меня где-то дальше будет dynamic_cast. Можно ли без этого? Наверняка. Умею ли я? Нет.

DimaTrushin commented on src/frame.h:
```diff
@@ -2,35 +2,33 @@
 private:
-    Screen& screen_;
+    Screen* screen_ = nullptr;
     std::vector<std::vector<sf::Color>> color_;
```
- Как я понял, это растровый экран =) Ждем рефакторинга.
- Вектор векторов -- плохая идея. Лучше уж заводить объект PixelScreen и внутри него держать один std::vector<double> размера height * width.

DimaTrushin commented on src/frame.h:
```diff
@@ -2,35 +2,33 @@
-    Screen& screen_;
+    Screen* screen_ = nullptr;
     std::vector<std::vector<sf::Color>> color_;
     std::vector<std::vector<double>> z_value_;
```
Вектор векторов -- надо заменить на один объект ZBuffer и внутри него один вектор, как выше про color_.

KiK0S replied:
Звучит как место для того, чтобы сделать шаблонный PixelScreen<double/sf::Color>. Думаю, что так и поступлю. Для контейнера шаблон звучит очень осмысленно

DimaTrushin commented on src/frame.cpp:
```diff
@@ -9,50 +9,50 @@
+            screen_->draw(line3d);
         }
-        SurfaceObject<double>* ptr = dynamic_cast<SurfaceObject<double>*>(current);
+        SurfaceObject* ptr = dynamic_cast<SurfaceObject*>(current);
```
- Неужели библиотека всерьез предлагает использование dynamic_cast в качестве решения??? Я бы этим говном никогда ни при каких условиях не пользовался, а если приходится пользоваться, то все равно бы не пользовался. Я понимаю, что бывают кривые библиотеки, но судя по всему это твоя ошибка, ты складываешь указатели на базовый класс, а потом кастуешь на реально хранимый класс, но при этом вперемешку хранишь объекты разного сорта. Это плохо, ненадежно и медленно в особенности dinamic_cast. А именно скорость отрисовки у тебя критическая.
- При этом я не вижу у тебя каста к каким-нибудь другим типам. А тогда не понятно, почему не хранить в контейнере указатель сразу на SurfaceObject?
- Если тебе надо хранить объекты разного сорта, то обычно для этого заводят разные контейнеры под каждый тип объекта свой контейнер.
ContainerForSurfaceObjects
ContainerForSomeOtherObjects
ContainerForNewSomeOtherObjects
- Если же у тебя нет такого разнообразия объектов, то вообще не понятно зачем такие касты нужны.
Кроме того, если уж ты хочешь в одном контейнере хранить указатели на базовый класс разных объектов, то тогда все это надо прокидывать через единый интерфейс с виртуальными методами, а не кастовать в run-time и обрабатывать каждый по-своему. Все что ты знаешь в compile-time надо делать в compile-time.
- Кроме того, если уж и делать правильный run-time polymorphism, то его лучше делать с value semantics, а не на указателях. Про это есть хороший доклад от Sean Parent
https://www.youtube.com/watch?v=QGcVXgEVMJg
Я на основе его доклада собрал шаблонный класс AnyObject, который поддерживает произвольный интерфейс. Если надо будет, я залью отдельно или дам тебе доступ к своим библиотекам.

- В сухом остатке, такие игры с dynamic_cast тебе тут совершенно не нужны. Это ошибка дизайна и уменьшение производительности, причем катострофическое.

KiK0S replied:
а, ты дошел до этого момента =)))) Сейчас прочитаю и осознаю, как делать лучше

KiK0S replied:
осознал, я могу делать wire-объекты как surface-наследников с нулевым числом поверхностей

DimaTrushin commented on src/frame.cpp:
```diff
@@ -9,50 +9,50 @@
             }
         }
     }
     std::vector<sf::Vertex> data;
```
Когда ты знаешь заранее размер вектора, которые тебе нужен, то я бы заранее преаллоцировал для него память, чтобы избежать лишних копирований. Можно добавить (указанного размера точно хватит)
data.reserve(screen_->kScreenSize * screen_->kScreenSize);
Вообще, можно выделить этот вектор в отдельный буффер внутри объекта.

DimaTrushin commented on src/matrix.h:
```diff
@@ -7,26 +7,41 @@
 
-    std::vector<double>& operator[] (size_t n);
-    const std::vector<double>& operator[] (size_t n) const ;
+    Cursor operator[] (size_t n);
```
Дерьмовая идея с прокси объектом в критической части кода. Это работает медленно, потому что требует дополнительного промежуточного объекта для обращения к элементу массива. Вместо этого надо перегрузить оператор скобки
double& operator()(size_t row, size_t column);
double operator()(size_t row, size_t column) const;
А в них имплементация такая
return data_[N * row + column];
Это еще сильнее ускорит доступ к элементам.

KiK0S replied:
Исправил

DimaTrushin commented on src/matrix.h:
```diff
@@ -7,26 +7,41 @@
+private:
+    struct Cursor;
+    struct ConstCursor;
+    static constexpr size_t N = 4;
```
Это тот уникальный случай, когда шаблон можно было бы оставить =) Потому что у тебя теперь это класс Matrix4d и не нужны никакие using-ги для него ниже.

KiK0S replied:
Да, я просто еще нейминги не поправил. Планирую оставить его в формате Matrix4d. Иначе мне кажется опять чем-то перегружено получится, хз

DimaTrushin commented on src/matrix.h:
```diff
@@ -7,26 +7,41 @@
 
-    template <size_t K>
-    Matrix<N, K> operator * (Matrix<M, K> other) const ;
+    Matrix operator * (Matrix other) const ;
```
Если уж по-хорошему писать класс матриц, то надо
1) Перегрузить все арифметические операторы.
2) Операторы должны быть двух сортов:
а) inplace
b) non-inplace
То есть для умножения у тебя должно быть
Matrix& operator*=(const Matrix& other) {
// умножаем текущую матрицу на other
return *this;
}
Это inplace метод. Теперь на его основе делаем non-inplace метод в виде friend operator
friend Matrix operator(const Matrix& first, const Matrix& second) {
Matrix resutl = first;
result *= second;
return result;
}
Тут еще и rvo затриггерится и нет code duplication.

DimaTrushin replied:
Не уверен, что в текущей сигнатуре передача by value -- хорошая идея. Я бы сделал хотя бы
Matrix operator * (const Matrix& other) const;
А вообще говоря переделал бы так, как написано в соседнем комментарии.

DimaTrushin commented on src/matrix.h:
```diff
@@ -7,26 +7,41 @@
 public:
     Matrix();
-    Matrix(std::vector<std::vector<double>>);
+    Matrix(std::array<double, N * N> data);
```
Очень плохой конструктор. Дело в том, что тут у тебя произойдет лишнее копирование массива. Когда ты пишешь
Matrix mat({1,2,3,....});
То  у тебя происходит aggregate initialization (https://en.cppreference.com/w/cpp/language/aggregate_initialization) для аргумента data в конструкторе. Инициализация inplce. Далее когда ты пишешь data_(data) то у тебя происходит копирование массива data в массив data_.
Вместо этого надо делать конструктор перхватывающий все 16 элементов матрицы напрямую. К сожалению структуры, которые aggregate initializable очень паршивы в этом смысле. Вот тут можно почитать, как эту грустную историю можно сделать менее грустной с помощью шаблонов
https://stackoverflow.com/questions/6893700/how-to-construct-stdarray-object-with-initializer-list
Тут ничего не поделаешь, другого способа нет.
Очень советую почитать про aggregate initialization. Я сам до сих пор не понимаю, как по-хорошему этой дрянью пользоваться без головной боли каждый раз.

KiK0S replied:
ну 16 элементов... :") может не страшно будет.

Я еще подумаю, что тут сделать, пока что не понял

DimaTrushin commented on src/matrix.cpp:
```diff
@@ -1,85 +1,87 @@
-const std::vector<double>& Matrix<N, M>::operator[] (size_t n) const {
+const double& Matrix::ConstCursor::operator[] (size_t n) const {
     if (n < 0 || n >= N) {
         throw std::out_of_range("bad index");
```
- Плохая идея кидать исключение, когда ты вышел за рамки массива. Есть две причины:
1) Все таки это ошибка программиста, которую он может решить, а не фейл системы. Потому тут лучше assert поставить
2) Наличие бросаемых исключений замедляет код. Исключения позволяют коду выйти из scope в любой точке, потому компилятор не может применять самые агрессивные методы оптимизации.
- Вообще, есть еще один совет. Во всех критичных для производительности методах ставить пометки noexcept. Тогда это максимально развязывает руки для компилятора в режиме оптимизации. Этот класс -- это один из таких. Тут ошибки могут быть только твои, потому пользуешься assert и все помечаешь noexcept включая операторы, функции доступа, модификации и т.д. Кроме конструкторов. (move конструкторы и assignment-ы автоматом сгенерируются noexcept).

KiK0S replied:
noexcept пока что не повесил, но на ассерты переписал

DimaTrushin commented on src/matrix.cpp:
```diff
@@ -1,85 +1,87 @@
-    result.w = pseudo_result[3][0];
+    std::array<double, N> tmp;
+    for (int i = 0; i < N; i++) {
+        tmp[i] = other.x * (*this)[i][0] + other.y * (*this)[i][1] + other.z * (*this)[i][2] + other.w * (*this)[i][3];
```
Когда внутри класса обращаешься к данным класса, можно непосредственно обращаться к данным
data_[N * i + 0]
и т.д.
Это ускорит процесс, избавит тебя от прокси объектов. Хотя если сделать оператор доступа как я сказал, это будет почти так же быстро, как и прямой доступ. И так как все определения будут в одной единице трансляции, то может быть компилятор даже заинлайнит код в этом случае.

KiK0S replied:
ну я просто оператор доступа переписал

DimaTrushin commented on src/matrix.cpp:
```diff
@@ -1,85 +1,87 @@
-template<size_t N, size_t M>
-Vector4d Matrix<N, M>::operator *(Vector4d other) const {
-    assert(M == 4);
+Vector4d Matrix::operator *(Vector4d other) const {
```
other немного сбивает с толку, я бы ожидал, что это будет другая матрица, а не вектор.

DimaTrushin commented on src/matrix.cpp:
```diff
@@ -1,85 +1,87 @@
+Matrix Matrix::inverse() const {
+    Matrix copy = (*this);
+    Matrix eye = identity_matrix();
+
```
Судя по всему ты тут делаешь Гаусса для нахождения обратной матрицы. Я бы выделил Гаусса в отдельный метод и тут собрал бы метод нахождения обратной, пользуясь Гауссом, а не в лоб на коленке.

KiK0S replied:
Эта штука сильно залезает во внутреннее устройство. Я бы выделил I II III тип операций, а вот дальше хз.

DimaTrushin replied:
Здесь подойдет любой вид рефакторинга. Вообще говоря я придерживаюсь такого правила: вложенные цикл = бан. Требуется перестроить немного мышление, чтобы понять, как с этим бороться. Обычно надо выделять в отдельные функции. Если при этом использовать инлайн функции, то читаемость возрастет, а производительность будет такой же.
Кроме того, посмотри на длину функции. Что в ней происходит  понять сходу не возможно. Надо сильно напрягаться, чтобы выделить хоть какие-то содержательные действия.

DimaTrushin commented on src/matrix.cpp:
```diff
@@ -1,85 +1,87 @@
-    }
+Matrix Matrix::inverse() const {
+    Matrix copy = (*this);
+    Matrix eye = identity_matrix();
```
eye плохое имя. Тут у тебя будет лежать результат вычислений. result куда более подходящее имя.

DimaTrushin commented on src/matrix.h:
```diff
@@ -7,26 +7,41 @@
+        int pos_;
+        const std::array<double, N * N>* data_;
+        ConstCursor(int pos, const std::array<double, N * N>* data);
+        const double& operator[] (size_t n) const ;
```
Базовые типы надо возвращать по значению, а не по константной ссылке. По ссылке получается дользе, потому что ссылка -- это адрес на double, а размер адреса и double одинаковый на 64 битных системах. Потому у тебя лишь лишнее разыменование адреса будет в этом случае. Что еще сильнее замедляет вычисления.

DimaTrushin reviewed (changes requested):
Я набросал еще несколько замечаний. Как обычно не смог пройтись по всему. И не оставлял аналогичные замечания. Есть несколько критичных предложений по оптимизации.

KiK0S commented:
Короче, я сейчас залил версию, которая опять работает, но работает лучше, чем раньше. Производительность я немножко поднял, но всякие косяки наверняка есть.

Чего тут нет пока что:
* Редизайна на 3 вспомогательных класса (World, Camera, Screen) и основной (Renderer)
* Нормальной отрисовки прямых и треугольников --- сейчас это самая тормозная часть

KiK0S committed:
restructurization

camera center moved to (size/2, size/2)

drawing line via pixels

add readme and gitignore

improved lines quality:

astyle one love

end file newlines

correct move

privacy modificators

sufrace objects

gif

some technical issues. deleted most of templates

git modules fix

new matrices

redesigned camera

small refactoring

enderer/screen/camera/world

move/rotate camera/world

pitch/roll/elevate/ so on interface

solid triangles

working triangle inner_point

reorganized min_y, but still a bug

fixed triangle order bug

new matrix constructor

optimize find_z

cube fix

smooth z-value, optimize draw triangle function

application file

save

unsaved changes

zet-buffer

new camera, need to fix

aaaaaa it works

paperwork

gitignore

KiK0S committed:
gitignore

DimaTrushin commented on test/test.cpp:
```diff
@@ -1,80 +1,89 @@
-        if (sf::Keyboard::isKeyPressed(sf::Keyboard::Escape)) {
-            screen.window_.close();
-            break;
+void setup(app::Application& application) {
```
Это выглядит как часть конструктора app::Application.

DimaTrushin commented on test/test.cpp:
```diff
@@ -1,80 +1,89 @@
+int main() {
+    app::Application application;
+    setup(application);
+    while (application.is_open()) {
```
Этот цикл выглядит как метод run() для Application.

DimaTrushin commented on test/test.cpp:
```diff
@@ -1,80 +1,89 @@
-        screen.update();
-        // std::this_thread::sleep_for(std::chrono::milliseconds(10));
     }
+    if (sf::Keyboard::isKeyPressed(sf::Keyboard::Left)) {
```
По хорошему такой перебор должен идти через switch или в худшем случае через else if, чтобы не делать лишние проверки.

Другой дизайн. Сделать функцию обрабатывающую в точности одно сообщение в очереди и делающее return после обработки и вынимания сообщения. А потом в цикле пока не опустеет ящик ее вызывать и обернуть это в функцию обработки всех текущих сообщений. Я бы сделал что-то вроде
```cpp
handle_all_events(...) {
  if (contains_close_event(...))
    return;
  while (!event_box_is_empty(...)) {
    handle_one_event(...);
  }
}
```
А в функции handle_one_event написать что-то вроде
```cpp
handle_one_event(event) {
  if(some_event(event)) {
    do_something();
    return;
  }
  ...
}
```
Или воспользоваться switch.

KiK0S replied:
Нет, потому что может быть нажато несколько кнопок одновременно - проверять надо все

DimaTrushin commented on src/application.cpp:
```diff
@@ -0,0 +1,92 @@
+namespace app {
+
+
+Application::Application(): world_(new World()), renderer_(new Renderer()) {}
```
Зачем тебе выделять по указателю World и Renderer?
Во-первых, если выделение памяти на World произошло, а Renderer кинул исключение, то деструктор не вызовется и память потечет. 
Во-вторых, вообще не понятно, зачем тебе тут указатели. Ты можешь просто сделать локальные объекты.
В-третьих, если все таки нужны указатели, то есть std::unique_ptr -- владеющий указатель, который безопасно работает с памятью.

DimaTrushin commented on src/application.cpp:
```diff
@@ -0,0 +1,92 @@
+
+
+bool Application::poll_event(sf::Event& event) {
+    return renderer_->poll_event(event);
```
Какого хрена у рендерера вообще есть такой метод?

DimaTrushin commented on src/application.cpp:
```diff
@@ -0,0 +1,92 @@
+}
+
+bool Application::is_open() const {
+    return renderer_->is_open();
```
Что это значит для рендерера?

DimaTrushin commented on src/application.cpp:
```diff
@@ -0,0 +1,92 @@
+}
+
+void Application::close() {
+    renderer_->close();
```
Что это значит для рендерера?

DimaTrushin commented on src/application.cpp:
```diff
@@ -0,0 +1,92 @@
+}
+
+void Application::roll(double angle) const {
+    renderer_->rotate_world(angle, 2);
```
Почему вращение мира идет через ренедер, если мир был отдельным объектом?? Какого хрена происходит? Это ошибка дизайна. У тебя очень путанная ответственность у объектов. Непонятно кто на кого ссылается, кто за что отвечает. По сути у тебя решение более путанное, чем просто набор глобальных переменных, потому что никто не уверен в том, кто про кого что знает. И от инкапсюляции нет никакого толку.

DimaTrushin replied:
Что за магическая константа 2? Если это указание Оси, то сделай enum class, в котором выдели константы с говорящими названиями вроде EAxes::OZ или что-то в этом духе.

DimaTrushin commented on src/camera.h:
```diff
@@ -1,30 +1,46 @@
+    const sf::Vector2f kDefaultPixel_ = sf::Vector2f(0, 0);
+
+    Point4d position_ = Point4d(0, 0, 0);
+    Point4d default_position_ = position_;
```
Почему дефолтная позиция строится по позиции, а не наоборот?
Почему дефолтная позиция не константа? и не constexpr вообще?
Это позиция фокусной точки или угла экрана или чего?

DimaTrushin commented on src/camera.h:
```diff
@@ -1,30 +1,46 @@
+
+    Point4d position_ = Point4d(0, 0, 0);
+    Point4d default_position_ = position_;
+    Matrix<3, 4> basis_screen_ = Matrix<3, 4>(std::vector<Point4d>({
```
тут лучше выделить статический приватный метод под
```cpp
static Matrix<3, 4> make_default_screen_basis() {
  return ...;
}
Matrix<3, 4> basis_screen_ = make_default_screen_basis();
```

DimaTrushin commented on src/camera.h:
```diff
@@ -1,30 +1,46 @@
     Matrix4d transform_ = Matrix4d::identity_matrix();
-    static constexpr double MAX_Z = 100000;
-    static constexpr double MIN_Z = -100000;
+    Matrix4d transform_camera_ = Matrix4d::identity_matrix();
```
В чем разница между двумя последними матрицами?

DimaTrushin commented on src/camera.cpp:
```diff
@@ -1,36 +1,91 @@
+}
+
+double Camera::get_max_z_value() const {
+    return 1000;
```
Почему эти константы не в объявлении класса? Они как-то коррелируют с константами, которые определены constexpr в классе?

DimaTrushin commented on src/camera.cpp:
```diff
@@ -1,36 +1,91 @@
+void Camera::create_transform() {
+    // Point4d pl = position_ + basis_screen_.row(0) * -1000 + basis_screen_.row(1) * -1000 + basis_screen_.row(2) * -1000;
+    // Point4d pr = position_ + basis_screen_.row(0) * 1000 + basis_screen_.row(1) * 1000 + basis_screen_.row(2) * 1000;
+    double l = -1000;
```
Что за магические константы. Они коррелируют с уже объявленными константами в классе?

DimaTrushin commented on src/camera.cpp:
```diff
@@ -1,36 +1,91 @@
+    double n = -1000;
+    double f = 1000;
+    Matrix4d canonical2Screen({
+        500.0, 0,     0, 499.5,
```
Магические константы.

DimaTrushin commented on src/camera.cpp:
```diff
@@ -1,36 +1,91 @@
+        0, 1, 0, - (b + t) / 2.0,
+        0, 0, 1, - (n + f) / 2.0,
+        0, 0, 0, 1});
+    Matrix4d magic({
```
Офигенное название.

KiK0S replied:
Самому нравится!

DimaTrushin commented on src/world.h:
```diff
@@ -0,0 +1,30 @@
+    std::vector<SurfaceObject*>::iterator end();
+
+private:
+    std::vector<SurfaceObject*> objects_;
```
Если здесь лежат указатели на интерфейсы, то кто владеет самими объектами за этими интерфейсами? Кто их хранит?

DimaTrushin commented on src/renderer.h:
```diff
@@ -0,0 +1,80 @@
+    \date Март 2015 года
+    \warning Данный класс создан только в учебных целях
+
+    Обычный дочерний класс, который отнаследован от ранее созданного класса Parent
```
Это что еще за дичь?

DimaTrushin commented on src/renderer.h:
```diff
@@ -0,0 +1,80 @@
+private:
+    constexpr static size_t kScreenSize_ = 1000;
+    sf::RenderWindow window_;
+    Screen * const screen_ = nullptr;
```
Какого хрена тут указатели на screen и camera? Почему ты не можешь передать в методы явно камеру и экран? Зачем их привязывать на уровне конструктора? Это же потенциальная underfined reference. Плюс неявные зависимости.

UPD. Опять увидел, что у тебя тут ненужные указатели. Почему ты не делаешь просто локальные переменные?

DimaTrushin commented on src/renderer.h:
```diff
@@ -0,0 +1,80 @@
+    sf::RenderWindow window_;
+    Screen * const screen_ = nullptr;
+    Camera * const camera_ = nullptr;
+    const static Vector4d kAxis_[3];
```
Почему не constexpr? Какое вообще назначение у этих констант?

DimaTrushin commented on src/renderer.cpp:
```diff
@@ -0,0 +1,226 @@
+    double min_y;
+    double max_y;
+    Matrix<2, 2> basis = triangle2d.create_basis();
+    for (int x = ceil(left_point.x); x <= floor(right_point.x); x++) {
```
Вот тут два вложенных цикла и три вложенных скоупа -- это ошибка дизайна. Рефактор.
Обычно тело внутри цикла зашивается в отдельную функцию с говорящим названием. Это делает код более читаемым.

DimaTrushin commented on src/renderer.cpp:
```diff
@@ -0,0 +1,226 @@
+
+void Renderer::draw(const Triangle4d& triangle4d) {
+    
+    Triangle2d triangle2d(camera_->project_point(triangle4d.a),
```
- Почему ты не сдела метод проектирования целиком треугольника? Это нарушение уровня абстракции.
- Что за треугольник ты тут строишь? У тебя имя объекта от имени типа не отличается.

DimaTrushin commented on src/renderer.cpp:
```diff
@@ -0,0 +1,226 @@
+    debug(camera_->transform_point(triangle4d.a));
+    debug("---");
+
+    if (triangle2d.a == triangle2d.b || triangle2d.c == triangle2d.a || triangle2d.c == triangle2d.b) {
```
Что за условие проверяется? Выдели это в отдельную функцию с говорящим названием.

DimaTrushin commented on src/renderer.h:
```diff
@@ -0,0 +1,80 @@
+    Matrix<2, 1> get_coords(int x, int y, Matrix<2, 2>& basis, sf::Vector2f& point) const ;
+private:
+    constexpr static size_t kScreenSize_ = 1000;
+    sf::RenderWindow window_;
```
Похоже на ошибку дизайна. Этот объект не должен быть внутри renderer. Как я понимаю именно этот объект отвечает за обработку событий экосистемы SFML.

DimaTrushin commented on src/renderer.cpp:
```diff
@@ -0,0 +1,226 @@
+    window_.draw(rectangle);
+}
+
+void Renderer::move_camera(Vector4d v) const {
```
В описании рендерера у него не указан функционал управления камерой. Это ошибка дизайна. Камера должна быть отдельным объектом, а renderer-у она передается в методах отрисовки явно.

Аналогично про экран.

DimaTrushin commented on src/renderer.cpp:
```diff
@@ -0,0 +1,226 @@
+const sf::Vector2f Renderer::kCenter_ = sf::Vector2f(Renderer::kScreenSize_ / 2, Renderer::kScreenSize_ / 2);
+const Vector4d Renderer::kAxis_[3] = {Vector4d(100, 0, 0), Vector4d(0, 100, 0), Vector4d(0, 0, 100)};
+
+Renderer::Renderer(): screen_(new Screen(this)), camera_(new Camera()),
```
Какого хрена экрану надо знать указатель на Renderer? Это скрытая зависимость и потенциально underfined reference, если Renderer не создан на куче или переместился.

DimaTrushin commented on src/screen.cpp:
```diff
@@ -1,98 +1,51 @@
-    delete camera_;
+Screen::Screen(Renderer* renderer): renderer_(renderer), z_value_(renderer_->get_screen_size()), color_(renderer_->get_screen_size()) {
+    assert(renderer_);
+    for (int i = 0; i  < renderer_->get_screen_size(); i++) {
```
Какого хрена? Почему нельзя просто создать экран по трем параметрам: width, height, z_depth?
Почему надо эти параметры тянуть из рендерера и еще помнить на него ссылку? Это проблема с mocing. Ты не сможешь тестировать такой объект, потому что у него есть скрытая зависимость, он не может существовать без Renderer-а.

DimaTrushin commented on src/screen.cpp:
```diff
@@ -1,98 +1,51 @@
-            int y = round(kek.y);
-            if (triangle2d.inner_point(x, y)) {
-                frame_->set_pixel(x, y, camera_->get_z_value(p), sf::Color::Black);
+    // TODO: calc non-zero pixels
```
Почему этот код выполняется не в методе Renderer, в который передается явно экран?

DimaTrushin commented on src/screen.cpp:
```diff
@@ -1,98 +1,51 @@
-        for (Vector4d& vertex : *wire_object) {
-            vertex = moving * vertex;
+void Screen::clear() {
+    for (int i = 0; i < renderer_->get_screen_size(); i++) {
```
Почму экран запрашивает размеры у рендерера и не помнит их сам? Что если текущие размеры экрана не совпадают с тем, что прилетеле из renderer-а? Это скрытые зависимости.

DimaTrushin commented on src/screen.cpp:
```diff
@@ -1,98 +1,51 @@
-    for (int i = 0; i < 3; i++) {
-        draw(Line4d(Vector4d(0, 0, 0), kAxis[i]));
+void Screen::set_pixel(int x, int y, double z, sf::Color color) {
+    if (x < 0 || y < 0 || x >= renderer_->get_screen_size() || y >= renderer_->get_screen_size() ||
```
- Выделить в отдельную функцию, првоеряющую корректность попадания пикселя на экран.
- Почему экран не сам помнит свои размеры?

DimaTrushin commented on CMakeLists.txt:
```diff
@@ -5,17 +5,16 @@ project(3d-drawer)
 add_subdirectory(src)
-add_subdirectory(SFML)
 
 SET(GCC_COVERAGE_COMPILE_FLAGS "-std=c++17")
```
gcc специфичный ключ. Формат стандарта надо не так подключать.
```cpp
set(CMAKE_CXX_STANDARD 17)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
```

KiK0S replied:
done

DimaTrushin reviewed (changes requested):
Я тут бегло пробежался глазами по коду и выделил несколько вещей, что попались на глаза сразу. Надо бы поправить.

KiK0S committed:
CMakeLists C++17

KiK0S committed:
zero-iteration doxygen docs

KiK0S committed:
drop pointers, just members for pipeline

KiK0S committed:
app logic moved from renderer to app

KiK0S committed:
camera moved out from renderer to app, refactor

KiK0S committed:
small fixes on statics const vars

KiK0S committed:
tmpcommit

KiK0S committed:
testing docker

KiK0S committed:
Github CI

KiK0S committed:
Update main.yml

KiK0S committed:
cube red lines

KiK0S committed:
working clipping

KiK0S committed:
speed up rendering triangles
